### PR TITLE
[clang] NFCI: use TemplateArgumentLoc for type-param DefaultArgument

### PR DESCRIPTION
**Pull request by mizvekov**
This is an enabler for a future patch.

This allows an type-parameter default argument to be set as an arbitrary TemplateArgument, not just a type.
This allows template parameter packs to have default arguments in the AST, even though the language proper doesn't support the syntax for it.

This will be used in a later patch which synthesizes template parameter lists with arbitrary default arguments taken from template specializations.

There are a few places we used SubsType, because we only had a type, now we use SubstTemplateArgument.
SubstTemplateArgument was missing arguments for setting Instantiation location and entity names.
Adding those is needed so we don't regress in diagnostics.

### CODE REVIEW

File: clang/include/clang/Sema/Sema.h [hunk 1]
```diff
@@ -10082,7 +10082,9 @@ class Sema final : public SemaBase {
 
   bool SubstTemplateArgument(const TemplateArgumentLoc &Input,
                              const MultiLevelTemplateArgumentList &TemplateArgs,
-                             TemplateArgumentLoc &Output);
+                             TemplateArgumentLoc &Output,
+                             SourceLocation Loc = {},
+                             const DeclarationName &Entity = {});
```
Comment (Endilll) [inline clang/include/clang/Sema/Sema.h:10087]: Changes to both declaration and implementation of `SubstTemplateArgument` doesn't seem related.

**Review by Endilll — COMMENTED** [review]
It looks like you have two sets of changes here:
1) the ones related to `TemplateTypeParmDecl::getDefaultArgument()`
2) the ones related to `Sema::SubstTemplateArgument()`

You don't seem to touch the latter in PR description. It would be nice if you can explain why you do both sets of changes in one PR, or split the PR in two.

File: clang/include/clang/Sema/Sema.h (hunk 1 above)
Comment (mizvekov) [reply clang/include/clang/Sema/Sema.h:10087]: There are a few places we used SubsType, because we only had a type, now we use SubstTemplateArgument.

SubstTemplateArgument was missing arguments for setting Instantiation location and entity names.
Adding those is needed so we don't regress in diagnostics.

Comment (Endilll) [reply clang/include/clang/Sema/Sema.h:10087]: Can you add this to the description?

**Review by Endilll — COMMENTED** [review]
LGTM, but you should wait for someone with more knowledge of our templates implementation.

**Review by erichkeane — APPROVED** [review]
I don't quite get the justification for this, but also don't see any downside for it, so I think this is acceptable.

**Comment by mizvekov** [comment]
> I don't quite get the justification for this, but also don't see any downside for it, so I think this is acceptable.

It's unfortunate GitHub does not show a PR stack like phab did.

This is needed for another patch you already reviewed: https://github.com/llvm/llvm-project/pull/92855

It allows forming default arguments for packs as part of TTP template argument deduction, based on the same principles already implemented in https://github.com/llvm/llvm-project/pull/89807
